### PR TITLE
Update new-backend docs

### PR DESCRIPTION
**Pull request by kodebach**
Incorporates the reviews from #4187

## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://master.libelektra.org/doc/contrib/documentation.md))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

File: src/libs/elektra/kdb.c
Comment (atmaxinger) [inline src/libs/elektra/kdb.c:641]:
```suggestion
		// addDupMounptoints duplicates everything, including reopening the plugins
```

File: doc/decisions/backend_plugin.md
Comment (markus2330) [inline doc/decisions/backend_plugin.md:12]:
```suggestion
- All existing plugins, except [hooks](hooks.md), should continue working as before.
```

File: doc/decisions/backend_plugin.md
Comment (markus2330) [inline doc/decisions/backend_plugin.md:24]:
```suggestion
- The current backend implementation was redeveloped into a backend plugin.
   The core of Elektra accesses backends through the standard plugin interface.
```

File: doc/decisions/backend_plugin.md
Comment (markus2330) [inline doc/decisions/backend_plugin.md:25]: Unclear if the "e.g." also refers to storage? If there are more exceptions than storage+resolver, it would be good to link there to the docu.

```suggestion
- The new backend plugin supports an unlimited number of plugins in any position where more than one plugin is sensible, e.g., unlimited plugins in `poststorage`, but only a single one in `storage` and `resolver`
```

File: doc/decisions/commit_function.md
Comment (markus2330) [inline doc/decisions/commit_function.md:1]: You can set this decision back to draft and add the options normally (not commented out). A related decision is if "returning the contract" should be its own function.

File: doc/dev/backend-plugins.md
Comment (markus2330) [inline doc/dev/backend-plugins.md:31]: :eyes:

File: doc/news/_preparation_next_release.md
Comment (markus2330) [inline doc/news/_preparation_next_release.md:55]: Please add more information here, this is also important for us to get a grasp what changed (not only for release).

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1596]:
```suggestion
 * @pre The KeySet @p returned must be a valid KeySet, i.e., constructed with ksNew().
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1598]:
```suggestion
 * @pre The KeySet @p returned must contain keys only from the `spec:/`, `dir:/`, `user:/`, `system:/`, `default:/` or `proc:/` namespaces.
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1599]:
```suggestion
 * @pre The @p parentKey Key must be a valid Key, i.e.., constructed with keyNew().
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1622]: To not make a constraint to `default:` keys added beforehand:

```suggestion
 * - `default:/` keys can be inserted by kdbGet() but they will always stem from a specification in `spec:/` keys
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:2194]: copy&paste errors

File: src/plugins/backend/README.md
Comment (markus2330) [inline src/plugins/backend/README.md:19]: :+1: please

File: src/plugins/missing/README.md
Comment (markus2330) [inline src/plugins/missing/README.md:12]: Can you please create an issue? We should get everything important to the issue tracker so that we know when we are finished and can release.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job! Keep up the good work! :heart:

File: src/libs/elektra/kdb.c
Comment (kodebach) [reply src/libs/elektra/kdb.c:1622]: > To not make a constraint to `default:` keys added beforehand

I accepted the suggestion, but not sure what you wanted to say here. I thought we want `default:/` keys to come from `spec:/` only (see https://github.com/ElektraInitiative/libelektra/pull/4484#discussion_r980332072). Pre-existing `default:/` keys will be removed, if the overlap with a loaded backend. Outside those backends, they will of course be kept, but that's the case for any other namespace too.

File: src/libs/elektra/kdb.c
Comment (kodebach) [inline src/libs/elektra/kdb.c:2193]:
```suggestion
 * @pre The KeySet @p returned must be a valid KeySet, e.g. constructed with ksNew().
```

File: src/libs/elektra/kdb.c
Comment (kodebach) [inline src/libs/elektra/kdb.c:2194]:
```suggestion
 * @pre The KeySet @p returend must only contain only keys in the `spec:/`,
```

File: src/plugins/missing/README.md
Comment (kodebach) [reply src/plugins/missing/README.md:12]: #4608

**Comment by kodebach** [comment]
@markus2330 Didn't have time yet to create a README for `backend`, but hopefully I addressed all the rest.

Regarding the decisions here:
- `commit_function.md`: This one is the important one that should be decided ASAP. I'm heavily in favour of not having separate functions (i.e. option 3 or 4). I would prefer the separate `phase` arguments, but that probably breaks to much stuff, so I'd go with option 3.
- `plugin_contract_function.md`: First of all, I know option 3 here, is not really part of this decision. But it didn't quite fit here and I just wanted to quickly write down the idea. I will remove it before the decision goes any further. This issue should also be decided soon. I prefer the `void ELEKTRA_PLUGIN_EXPORT (KeySet * ks)` variant, because of all the other things it makes possible.
- `plugin_struct.md`: That's basically a list of ideas I had while looking into the other two decisions. In short, I think there is **a lot** of potential to slim down the `struct _Plugin` and improve some semi-related things. Please, ignore this file for now. I just wanted to write down the ideas.

> **Note** I won't do any more work on these decisions in this PR. Feel free to comment, but I will create separate PRs for each of them once this PR is merged.

File: src/libs/elektra/kdb.c
Comment (markus2330) [reply src/libs/elektra/kdb.c]: The change is good then. But it would be good to have these two sentences in the docu, too.

**Comment by markus2330** [comment]
> Didn't have time yet to create a README for backend,

Would be good to have it!

> but hopefully I addressed all the rest.

What about the other reviews? Are they also incorporated here?

> I will create separate PRs for each of them once this PR is merged.

Excellent idea!

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1598]: analog changes also with Key:

```suggestion
 * @pre The Key @p parentKey must be a valid Key, i.e., constructed with keyNew().
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1613]: What we do internally or not should not be API docu. Please only describe user-visible behavior.

```suggestion
 * @warning If you later call kdbSet() with the same @p handle you must make sure to
 *          pass all keys from @p ks, which you do not want to remove.
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1621]:
```suggestion
 * - `default:/` keys can be inserted by kdbGet() but they will always stem from a specification in `spec:/` keys.
```

**Comment by kodebach** [comment]
> What about the other reviews? Are they also incorporated here?

Yes, everything from #4187 should be addressed.

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1631]: - Please use newlines as described in documentation guidelines for new sentences.
- Please add misspellings of namespaces to scripts/sed
- What is loadable? Please use consistent terminology (e.g. persistent)

```suggestion
 * @note If @p parentKey is in the cascading namespace, keys of all loadable namespaces (see above) will be loaded. 
 *       This is generally the recommended approach.
```

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:2192]:
```suggestion
 * @pre The KeySet @p returned must be a valid KeySet, i.e., constructed with ksNew().
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Still a few little things open. @lawli3t can you take a look at the API docu, too?

File: src/libs/elektra/kdb.c
Comment (kodebach) [reply src/libs/elektra/kdb.c:1613]: IIRC this was taken almost 1:1 from the pre-`new-backend` docs, but sure I'll change it

File: src/libs/elektra/kdb.c
Comment (kodebach) [reply src/libs/elektra/kdb.c:1631]: > Please use newlines as described in documentation guidelines for new sentences.

I generally agree that every sentence should be on a separate line in Markdown files and other purely "docs text" files. However, in code files many (if not most) people will not have line wrapping enabled (for very good reasons), so putting each sentence on separate line in doxygen comments just makes it needlessly hard to read in the file. It also makes the file kind of ugly. All the C line will be automatically wrapped and reformatted at a certain length via `clang-format`. But doxygen comments will stick out and may go far beyond that limit.

In this specific case, it is probably still an acceptable line length, if we keep the sentence in one line, so accepted the change.

> What is loadable? Please use consistent terminology (e.g. persistent)

Do you need to say "See 16 lines above" and update it all the time???? The sentence already says "see above" and there is literally a section called "Loadable namespaces" above. Sometimes I truly think you purposefully ignore things.

File: src/libs/elektra/kdb.c
Comment (markus2330) [inline src/libs/elektra/kdb.c:1599]: Not done for following lines.

Comment (kodebach) [reply src/libs/elektra/kdb.c:1599]: You only posted a suggestion for one line. I just accepted it.

Comment (markus2330) [reply src/libs/elektra/kdb.c:1599]: This I saw. There was text above it.

Comment (kodebach) [reply src/libs/elektra/kdb.c:1599]: Sorry, wasn't meant to put blame on you. It was just an explanation of what happened. I should've checked if there are similar issues nearby

**Review by markus2330 — APPROVED** [review]
LGTM, many improvements. Decisions will be handled later anyway. Please rebase.

**Comment by kodebach** [comment]
@markus2330 rebased, please merge when CI completes

**Comment by markus2330** [comment]
Please don't change the code in this already-reviewed documentation PR.

**Comment by kodebach** [comment]
> Please don't change the code in this already-reviewed documentation PR.

It's my PR, please don't tell me what to do with it...

I intentionally, split https://github.com/ElektraInitiative/libelektra/pull/4573/commits/85e8e19edcbfdb41fded4f8204a012cc6c162765 and https://github.com/ElektraInitiative/libelektra/pull/4573/commits/3694bb885fe04c638a59f5716db9c1124e52b60b.

The first one was a clear bug and I don't think there can be any objection to the fix. It even matches, what the docs already say. The second one does add new functionality, whether or not we want to add it is up for debate. But that's why I requested a new review. Well now I did... You commented faster than I could request the review...

**Review by markus2330 — CHANGES_REQUESTED** [review]
Please split documentation and code in this PR. In the current form I cannot review it anymore.

**Comment by kodebach** [comment]
@markus2330 I removed the last commit from the PR. Now the only new commit since your last review is [this](https://github.com/ElektraInitiative/libelektra/pull/4573/commits/85e8e19edcbfdb41fded4f8204a012cc6c162765). I believe you should be able to review that. It is a very small fix to make `kdbGet` do what the docs already claimed and adds a new test to check for the behaviour. It's also not new behaviour, but restores behaviour that existed before the new-backend merge (it was accidentally removed and the tests changed).

**Comment by kodebach** [comment]
I have now also remove the minor fix for "kdbGet never returns 0", because for some unexplained reason the test didn't work for the FULL build. I believe you can now just merge this PR, when the CI is done. There are now changes since your last review.

**Comment by markus2330** [comment]
jenkins build libelektra please

**Comment by markus2330** [comment]
Great work, thank you! :revolving_hearts:

**Pull request merged**
